crates/task-impls/src/consensus.rs

```diff
@@ -392,31 +384,65 @@ impl<TYPES: NodeType, I: NodeImplementation<TYPES>, A: ConsensusApi<TYPES, I> +
-                    if let Some(timeout_cert) = proposal.data.timeout_certificate.clone() {
-                        if timeout_cert.get_data().view != view - 1 {
-                            warn!("Timeout certificate for view {} was not for the immediately preceding view", *view);
-                            return;
-                        }
+                    if let Some(proposal_cert) = &self.proposal_cert {
+                        match proposal_cert {
+                            ViewChangeEvidence::Timeout(timeout_cert) => {
+                                if timeout_cert.get_data().view != view - 1 {
+                                    warn!("Internal timeout certificate for view {} was not for the immediately preceding view", *view);
+                                    return;
+                                }
 
-                        if !timeout_cert.is_valid_cert(self.timeout_membership.as_ref()) {
-                            warn!("Timeout certificate for view {} was invalid", *view);
-                            return;
-                        }
-                    } else if let Some(cert) = &self.view_sync_cert {
-                        // View sync certs _must_ be for the current view.
-                        if cert.view_number != view {
-                            debug!(
+                                if !timeout_cert.is_valid_cert(self.timeout_membership.as_ref()) {
+                                    warn!(
+                                        "Internal timeout certificate for view {} was invalid",
+                                        *view
+                                    );
+                                    return;
+                                }
+                            }
+                            ViewChangeEvidence::ViewSync(view_sync_cert) => {
+                                // View sync certs _must_ be for the current view.
+                                if view_sync_cert.view_number != view {
+                                    debug!(
                                 "Cert view number {:?} does not match proposal view number {:?}",
-                                cert.view_number, view
+                                view_sync_cert.view_number, view
                             );
-                            return;
+                                    return;
+                                }
+
+                                // View sync certs must also be valid.
+                                if !view_sync_cert.is_valid_cert(self.quorum_membership.as_ref()) {
+                                    debug!("Invalid ViewSyncFinalize cert provided");
+                                    return;
+                                }
+                            }
                         }
+                    }
+                    if let Some(received_proposal_cert) = proposal.data.proposal_certificate.clone()
```

Comment:
@bfish713 This was the issue. A received timeout cert was not being handled and the code assumed it was somehow already stored. So I added this additional check back. Please advise if assigning at the end of this block is correct. I believe that it should be done since otherwise we lose the timeout cert in the proposal function.

Reply:
Hmm.. well that fixed 4 of the failures. Something else is going on.

Reply:
Got it. The logic was checking a stale timeout cert.

Reply:
I'm confused by the logic above where we are matching against our cert?  That seems wrong.  We just received a proposal from someone else, we should verify that it has the appropriate evidence which should only involve checking whatever is on the proposal we just got.  `self.propsal` cert should only be for creating our own proposals.  It should only be set when we formed a timeout certificate, or a got a view sync finalized certificate.

Reply:
in other words I think in the `QuorumProposalRevc` event handler we should only do
```rust
if proposal.data.justify_qc.get_view_number() != view - 1 {
     if let Some(received_proposal_cert) = proposal.data.proposal_certificate.clone() {
        match received_proposal_cert {
            ViewChangeEvidence::Timeout(timeout_cert) => { // is it a valid TC for proposal.view - 1?}
            ViewChangeEvidence::ViewSync(view_sync_cert) => {  // is it a valid ViewSyncCert for proposal.view - 1 }
```